src/utils/getProjectConfig.ts

```diff
@@ -4,32 +4,128 @@ import {existsSync, readFileSync} from 'fs';
+const CONFIG_FILE_PATH_SERVICE = './services.config.json';
+
+/**
+ * Given name for the projects local configuration file needed for our operations.
+ */
+const CONFIG_FILE_PATH_PROJECT = './project.config.json';
 
-let config: ConfigFile|null = null;
+let config: Config;
 
 /**
  * Returns the config json for the extending project.
  */
-export function getProjectConfig(): ConfigFile {
+export function getProjectConfig<T extends Config = Config>(): Readonly<T> {
     if (config) {
-        return config;
+        return config as T;
     }
 
     // Check that the file exists locally
-    if (!existsSync(CONFIG_FILE_PATH)) {
-        throw new Error(`Config file not found in project root.  Please check ${CONFIG_FILE_PATH} exists.`);
+    let configPath: string;
+    let isService: boolean;
+    if (existsSync(CONFIG_FILE_PATH_SERVICE)) {
+        configPath = CONFIG_FILE_PATH_SERVICE;
+        isService = true;
+    } else if (existsSync(CONFIG_FILE_PATH_PROJECT)) {
+        configPath = CONFIG_FILE_PATH_PROJECT;
+        isService = false;
+    } else {
+        throw new Error(`Config file not found in project root.  Please check either ${CONFIG_FILE_PATH_SERVICE} or ${CONFIG_FILE_PATH_PROJECT} exists.`);
+    }
+
+    // Parse config
+    try {
+        config = JSON.parse(readFileSync(configPath, 'utf8'));
+    } catch (e) {
+        throw new Error(`Error parsing ${configPath}, check JSON is valid`);
+    }
+
+    // Check for required properties
+    const missingProperties = ['NAME', 'TITLE', 'PORT', 'CDN_LOCATION'].filter((prop) => {
+        return !config.hasOwnProperty(prop);
+    });
+    if (missingProperties.length > 0) {
+        throw new Error(`Couldn't find one or more required properties in config file: ${missingProperties.join(', ')} (${configPath})`);
+    }
+
+    // Apply any user-specific overrides
+    const userConfigPath: string = configPath.replace('.config.', '.user.');
+    if (existsSync(userConfigPath)) {
+        let userConfig;
+        try {
+            userConfig = JSON.parse(readFileSync(userConfigPath, 'utf8'));
+        } catch (e) {
+            throw new Error(`Error parsing ${userConfigPath}, check JSON is valid`);
+        }
+
+        config = {...config, ...userConfig};
+    }
+
+    // Apply CLI/env overrides
+    const {env} = process;
+    const argList = Object.keys(config) as (keyof Config)[];
+    argList.forEach(<K extends keyof Config>(key: K) => {
+        if (env.hasOwnProperty(key)) {
+            console.log(`Using ${key}:'${env[key]}' from environment vars`);
+
+            // All parameters coming from 'env' will be strings, need to parse as correct type.
+            // Will use the type of the default value to decide how to parse.
+            config[key] = parseCLIArg(env[key]!, config[key]);
+        }
+    });
+
+    config.IS_SERVICE = isService;
+    return config as T;
+}
+
+function parseCLIArg<T>(input: string, defaultValue: T) {
```

Comment:
Return type on function

Reply:
Requires some `as unknown as <type>` casts, due to the runtime type-checking that goes on in here.